daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -642,94 +643,171 @@ private[lf] object Speedy {
+                value match {
+                  case V.ValueContractId(cid) =>
+                    addGlobalCid(cid)
+                    SValue.SContractId(cid)
+                  case V.ValueNumeric(d) =>
+                    elemType match {
+                      case TNat(s) =>
+                        data.Numeric.fromBigDecimal(s, d).fold(crash, SValue.SNumeric)
+                      case _ =>
+                        typeMismatch
+                    }
+                  case V.ValueOptional(mb) =>
+                    mb match {
+                      case Some(value) => SValue.SOptional(Some(go(elemType, value)))
+                      case None => SValue.SValue.None
+                    }
+                  // list
+                  case V.ValueList(ls) =>
+                    SValue.SList(ls.map(go(elemType, _)))
+
+                  // textMap
+                  case V.ValueTextMap(entries) =>
+                    SValue.SGenMap(
+                      isTextMap = true,
+                      entries = entries.iterator.map { case (k, v) =>
+                        SValue.SText(k) -> go(elemType, v)
+                      },
+                    )
                   case _ =>
-                    crash(s"definition for variant $id not found")
+                    typeMismatch
                 }
-              case None =>
-                throw SpeedyHungry(
-                  SResultNeedPackage(
-                    id.packageId,
-                    pkg => {
-                      compiledPackages = pkg
-                      returnValue = go(value)
-                    },
-                  )
-                )
+              case keyType :: valueType :: Nil =>
+                value match {
+                  // genMap
+                  case V.ValueGenMap(entries) =>
+                    SValue.SGenMap(
+                      isTextMap = false,
+                      entries = entries.iterator.map { case (k, v) =>
+                        go(keyType, k) -> go(valueType, v)
+                      },
+                    )
+                  case _ =>
+                    typeMismatch
+                }
+              case _ =>
+                typeMismatch
+            }
+          case TTyCon(tyCon) =>
+            val signature = compiledPackages.signatures(tyCon.packageId)
+            value match {
+              case V.ValueRecord(_, fields) =>
+                signature.lookupRecord(tyCon.qualifiedName) match {
+                  case Right((params, DataRecord(fieldsDef))) =>
+                    lazy val subst = (params.toSeq.view.map(_._1) zip tyArgs).toMap
+                    val n = fields.length
+                    var i = 0
+                    val values = new util.ArrayList[SValue](n)
+                    while (i < n) {
+                      val typDef = fieldsDef(i)._2
+                      val typ = if (tyArgs.isEmpty) {
+                        // optimization
+                        typDef
```

Comment:
have you considered adding this to `replaceParameters` instead? Seems a bit more robust to get this right once in the callee than trying to get it right at every call site.